phare/pp/diagnostic/data/particle_level_overlap.py

```diff
@@ -0,0 +1,137 @@
+from phare.pp.diagnostics import Diagnostic, Patch, _ParticlePatchData, Particles
+from .particle_overlap import ParticleOverlap, get_ghost_patch
+from .particle_patch_overlap import DomainParticleOverlap
+from .overlap import Overlap
+
+from phare.pp.diagnostic.patch_data import particlesForDiags
+from phare.pp.diagnostic.patch import REFINEMENT_RATIO
+
+
+class LevelParticleOverlap(ParticleOverlap):
+    def __init__(
+        self,
+        coarseDomainPatch,
+        fineLevelGhostPatch,
+        data_name,
+        nGhosts,
+        sizes,
+        coarsePatchGhostPatch,
+    ):
+        ParticleOverlap.__init__(
+            self, coarseDomainPatch, fineLevelGhostPatch, data_name, nGhosts, sizes
+        )
+        self.coarseDomainPatch = coarseDomainPatch
+        self.coarsePatchGhostPatch = coarsePatchGhostPatch
+        self.fineLevelGhostPatch = fineLevelGhostPatch
+
+
+def getLevelGhostOverlaps(diags):
+    if isinstance(diags, dict) and _ParticlePatchData.__name__ in diags:
+        diags = diags[_ParticlePatchData.__name__]
+
+    # Merge diagnotics of same population into single Particles object
+    if all([isinstance(diag, Diagnostic) for diag in diags]):
+        diags = particlesForDiags(diags)
+
+    assert all([isinstance(diag, Particles) for diag in diags])
+
+    return _levelGhost(LevelParticleOverlap, diags)
+
+
+def _levelGhost(OverlapType, diags):
+    overlaps = []
+    for particles in diags:
+        for refineLevel in range(1, len(list(particles.domainDiag.levels.keys()))):
+            coarseDomainPatches = list(
+                particles.domainDiag.levels[refineLevel - 1].patches.values()
+            )
+            for coarseDomain in coarseDomainPatches:
+                fineLevelGhostPatches = list(
+                    particles.lGhostDiag.levels[refineLevel].patches.values()
+                )
+                for fineLevelGhost in fineLevelGhostPatches:
+                    overlaps.extend(
+                        _intersect_coarser_1d(
+                            OverlapType,
+                            coarseDomain,
+                            get_ghost_patch(
+                                particles, coarseDomain, DomainParticleOverlap
+                            ),
+                            fineLevelGhost,
+                        )
+                    )
+
+    return overlaps
+
+
+def _intersect_coarser_1d(OverlapType, coarseDomain, coarsePatchGhost, fineLevelGhost):
+    assert all(
+        [isinstance(x, Patch) for x in [coarseDomain, coarsePatchGhost, fineLevelGhost]]
+    )
+
+    return _calculate_coarser_1d(
+        OverlapType, coarseDomain, coarsePatchGhost, fineLevelGhost, "particles"
+    )
+
+
+def _calculate_coarser_1d(
```

Comment:
I don't understand what this function does.
It seems to return overlaps, I don't get why so much code is needed for that, and why looking at particle iCell is also needed for that.
Could you please explain me?

Reply:
The iCell check is there to show, there is actually a level ghost particle overlap, if there is no iCell, it's probably that there's a patch ghost particle overlap so there's no levelghost data.

Reply:
I don't get it... :(

Reply:
levelghost data does not exist if on the fine level there is a patchghost overlap
so even if we find a geometrical overlap, it's only "a possible" overlap, and if there is no icell value exist, then this tells us it is "hidden" by a patchghost overlap